pkg/lightning/restore/check_info.go

```diff
@@ -0,0 +1,501 @@
+			hasHeader := rc.cfg.Mydumper.CSV.Header
+			parser = md.NewCSVParser(&rc.cfg.Mydumper.CSV, reader, blockBufSize, rc.ioWorkers, hasHeader)
+		case md.SourceTypeSQL:
+			parser = md.NewChunkParser(rc.cfg.TiDB.SQLMode, reader, blockBufSize, rc.ioWorkers)
+		case md.SourceTypeParquet:
+			parser, err = md.NewParquetParser(ctx, rc.store, reader, dataFileMeta.Path)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		default:
+			msgs = append(msgs, fmt.Sprintf("file '%s' with unknown source type '%s'", dataFileMeta.Path, dataFileMeta.Type.String()))
+		}
+		err = parser.ReadRow()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		colsFromDataFile := parser.Columns()
+		colCountFromDataFile := len(parser.LastRow().Row)
+
+		colCountFromTiDB := len(info.Core.Columns)
+		colsFromTiDB := make([]string, 0, colCountFromTiDB)
+		defaultCols := make(map[string]struct{})
+		for _, col := range info.Core.Columns {
+			if hasDefault(col) || (info.Core.ContainsAutoRandomBits() && mysql.HasPriKeyFlag(col.Flag)) {
+				// this column has default value or it's auto random id, so we can ignore it
+				defaultCols[col.Name.L] = struct{}{}
+			}
+			colsFromTiDB = append(colsFromTiDB, col.Name.L)
+		}
+		// tidb_rowid have a default value.
+		defaultCols[model.ExtraHandleName.String()] = struct{}{}
+
+		if colsFromDataFile == nil {
+			// when there is no columns name in data file. we must insert data in order.
+			// so the last several columns either can be ignored or has a default value.
+			for i := colCountFromDataFile; i < colCountFromTiDB; i++ {
+				if _, ok := defaultCols[colsFromTiDB[i]]; !ok {
+					msgs = append(msgs, fmt.Sprintf("TiDB schema `%s`.`%s` has %d columns,"+
+						"and data file has %d columns, but column %s are missing the default value,"+
+						"please give column a default value to skip this check",
+						tableInfo.DB, tableInfo.Name, colCountFromTiDB, colCountFromDataFile, colsFromTiDB[i]))
+				}
+			}
+		} else {
+			// compare column names and make sure
+			// 1. TiDB table info has data file's all columns(besides ignore columns)
+			// 2. Those columns not introduced in data file always have a default value.
+			colMap := make(map[string]struct{})
+			for col := range igCols {
+				colMap[col] = struct{}{}
+			}
+			for _, col := range colsFromTiDB {
+				colMap[col] = struct{}{}
+			}
+			// tidb_rowid can be ignored in check
+			colMap[model.ExtraHandleName.String()] = struct{}{}
+			for _, col := range colsFromDataFile {
+				if _, ok := colMap[col]; !ok {
+					msgs = append(msgs, fmt.Sprintf("TiDB schema `%s`.`%s` doesn't have column %s, "+
+						"please check table schema or use tables.ignoreColumns to ignore %s",
+						tableInfo.DB, tableInfo.Name, col, col))
+				} else {
+					// remove column for next iteration
+					delete(colMap, col)
+				}
+			}
+			// if theses rest columns don't have a default value.
+			for col := range colMap {
+				if _, ok := defaultCols[col]; ok {
```

Comment:
If the column in ignoreColumns don't exists in source file and table schema, it will cause false positive check. It's better to report a different warning

Reply:
`If the column in ignoreColumns don't exist in the  source file`
=> user set the wrong config of `ignoreColumns`
=> this is not false positive check
=> the message `please give a default value for %s or **add this column in data file**`